sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,316 @@
+   * Get a Cassandra cluster using hosts and port.
+   */
+  private Cluster getCluster(List<String> hosts, int port) {
+    return Cluster.builder()
+        .addContactPoints(hosts.toArray(new String[0]))
+        .withPort(port)
+        .build();
+  }
+
+  /**
+   * Gets the list of token ranges that a table occupies on a give Cassandra node.
+   *
+   * <p>NB: This method is compatible with Cassandra 2.1.5 and greater.
+   */
+  private static List<TokenRange> getTokenRanges(
+      Cluster cluster, String keyspace, String table) {
+    try (Session session = cluster.newSession()) {
+      ResultSet resultSet =
+          session.execute(
+              "SELECT range_start, range_end, partitions_count, mean_partition_size FROM "
+                  + "system.size_estimates WHERE keyspace_name = ? AND table_name = ?",
+              keyspace,
+              table);
+
+      ArrayList<TokenRange> tokenRanges = new ArrayList<>();
+      for (Row row : resultSet) {
+        TokenRange tokenRange =
+            new TokenRange(
+                row.getLong("partitions_count"),
+                row.getLong("mean_partition_size"),
+                row.getLong("range_start"),
+                row.getLong("range_end"));
+        tokenRanges.add(tokenRange);
+      }
+      // The table may not contain the estimates yet
+      // or have partitions_count and mean_partition_size fields = 0
+      // if the data was just inserted and the amount of data in the table was small.
+      // This is very common situation during tests,
+      // when we insert a few rows and immediately query them.
+      // However, for tiny data sets the lack of size estimates is not a problem at all,
+      // because we don't want to split tiny data anyways.
+      // Therefore, we're not issuing a warning if the result set was empty
+      // or mean_partition_size and partitions_count = 0.
+      return tokenRanges;
+    } catch (InvalidQueryException e) {
+      LOG.warn("Failed to fetch size estimates for {}.{} from system.size_estimates table. "
+          + "The number of created bundles may be inaccurate. Please make sure you use "
+          + "Cassandra 2.1.5 or newer.", keyspace, table, e);
+      return Arrays.asList(
+          new TokenRange(0L, 0L, 0L, 0L));
+    }
+  }
+
+  /**
+   * Compute the percentage of token addressed compared with the whole tokens in the cluster.
+   */
+  @VisibleForTesting
+  protected static float getRingFraction(List<TokenRange> tokenRanges) {
+    long addressedTokens = 0L;
+    for (TokenRange tokenRange : tokenRanges) {
+      addressedTokens = addressedTokens + distance(tokenRange.rangeStart, tokenRange.rangeEnd);
+    }
+    return addressedTokens / TOTAL_TOKENS;
+  }
+
+  /**
+   * Measure distance between two tokens.
+   */
+  @VisibleForTesting
+  protected static long distance(long token1, long token2) {
+    if (token2 > token1) {
+      return token2 - token1;
+    } else {
+      return token2 - token1 + TOTAL_TOKENS;
```

Comment:
Since the value of TOTAL_TOKENS is currently meaningless, that means the unit tests never hit this branch, because it would have returned an incorrect value. Indeed, I see only one unit test for distance and it covers only the if branch, but not the else branch. Please add a test covering this one too.